packages/legend-query/src/stores/QueryBuilderValueSpecificationCloneHelper.ts

```diff
@@ -0,0 +1,196 @@
+/**
+ * Copyright (c) 2020-present, Goldman Sachs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  AbstractPropertyExpression,
+  AlloySerializationConfigInstanceValue,
+  CollectionInstanceValue,
+  EnumValueInstanceValue,
+  FunctionExpression,
+  GenericType,
+  GenericTypeExplicitReference,
+  InstanceValue,
+  INTERNAL__PropagatedValue,
+  INTERNAL__UnknownValueSpecification,
+  LambdaFunctionInstanceValue,
+  MappingInstanceValue,
+  Multiplicity,
+  PairInstanceValue,
+  PrimitiveInstanceValue,
+  PRIMITIVE_TYPE,
+  PropertyGraphFetchTreeInstanceValue,
+  PureListInstanceValue,
+  PureModel,
+  RootGraphFetchTreeInstanceValue,
+  RuntimeInstanceValue,
+  SimpleFunctionExpression,
+  TYPICAL_MULTIPLICITY_TYPE,
+  ValueSpecification,
+  ValueSpecificationVisitor,
+  VariableExpression,
+} from '@finos/legend-graph';
+import { UnsupportedOperationError } from '@finos/legend-shared';
+
+export class QueryBuilderValueSpecificionCloneHelper
```

Comment:
This is fair, but we don't need to do a full clone right? If we do, we'd better just turn it into `V1_protocol` and then build it back rather than having to do this

Reply:
As you said, there's a loophole here with the roundtrip approach since we don't deal with `INTERNAL__PropagatedValue` there. But I still think we should investigate if there's a less overkill way to clone.

Reply:
We can use a while loop to walk the property expression chain. Remember, we can also rewrite a visitor to a flat method :D, especially when we default out most of the methods of the visitor. Also, we got a bug here, since we have to deal with `subtyping`, i.e. the `SimpleFunctionExpression` case. So my take is we really have to do the `while` loop to walk the property expression chain.